### PR TITLE
Merge CachedRenderPipelinePhaseItem and PhaseItem

### LINKED ISSUE #7159: Consider Merging CachedRenderPipelinePhaseItem and PhaseItem
## What problem does this solve or what need does it fill?

I believe that every `PhaseItem` requires a `RenderPipeline` to be drawn.
All of our render pipelines are cached in the `PipelineCache`. 
Thus I think that it would make sense to merge the `CachedRenderPipelinePhaseItem` trait and `PhaseItem`.
This makes implementing new phase items simpler and more straightforward.

```rust
impl PhaseItem for Opaque3d {
    type SortKey = Reverse<FloatOrd>;

    #[inline]
    fn entity(&self) -> Entity {
        self.entity
    }

    #[inline]
    fn sort_key(&self) -> Self::SortKey {
        Reverse(FloatOrd(self.distance))
    }

    #[inline]
    fn draw_function(&self) -> DrawFunctionId {
        self.draw_function
    }

    #[inline]
    fn sort(items: &mut [Self]) {
        // Key negated to match reversed SortKey ordering
        radsort::sort_by_key(items, |item| -item.distance);
    }

    #[inline]
    fn cached_pipeline(&self) -> CachedRenderPipelineId {
        self.pipeline
    }
}
```

Do you know of a use case that does not require a render pipeline, or that does not use the pipeline cache?
Is there another reason for them to stay separated?

## What solution would you like?

If there are no objections, then I can create a PR implementing this change.

## What alternative(s) have you considered?

Leave everything as is (e.g. separate implementations for `PhaseItem` and `CachedRenderPipelinePhaseItem`).

### PR DESCRIPTION
**Pull request by kurtkuehnert**
# Objective

- Fixes #7159

See the reasons for this PR in the issue mentioned above.
I am still not sure whether this covers all use cases. 
Let me know if you find a reason to let them stay separated.

## Solution

- removes `CachedRenderPipelinePhaseItem` and adds its method `render_pipeline`  to `PhaseItem` instead

---

## Changelog

- removed `CachedRenderPipelinePhaseItem`
- added `pipeline_id` method to `PhaseItem` replacing `CachedRenderPipelinePhaseItem::render_pipeline`
- renamed the field `pipeline` to `pipeline_id` on the types `Transparent2d`, `Opaque3d`, `AlphaMask3d`, `Transparent3d`, and `Shadow`

## Migration Guide

- change the occurrences of the `pipeline` field on the phase items: `Transparent2d`, `Opaque3d`, `AlphaMask3d`, `Transparent3d`, and `Shadow` to `pipeline_id`
- instead of separately implementing `CachedRenderPipelinePhaseItem` for your phase items, now use the `PhaseItem::pipeline_id` method

### CODE REVIEW

**Review by james7132 — APPROVED** [review]
LGTM in terms of implementation. Can't say that about the open question of whether we always need a CachedRenderPipelineId though.

**Comment by kurtkuehnert** [comment]
Yeah, I am unsure as well. I think that we should encourage using our abstractions. Thus we should make using them as easy and brief as possible. On the other hand, there might exist use cases where using only the `RenderPhase`  without the `PipelineCache` is favorable. Assigning one and only one pipeline per phase item seems intuitive though.

I think that this change is controversial and @cart should take a look at it.

File: crates/bevy_ecs/src/schedule_v3/executor/multi_threaded.rs
Comment (kurtkuehnert) [inline crates/bevy_ecs/src/schedule_v3/executor/multi_threaded.rs:188]: Somehow the code from bevy main did not compile. This fixed it. What is going on?

**Comment by kurtkuehnert** [comment]
Hmm, the CI failures do not seem to be caused by this PR.

**Comment by superdump** [comment]
bors try

**Comment by superdump** [comment]
I guess using a `CachedRenderPipelineId` means looking up the pipeline from a `Vec`, which has some small overhead compared to just passing the pipeline directly. However, I think it's only simple cases where there would be a low frequency of lookups per frame where the pipeline could be stored in a way that it can be used directly. I'm not sure there's a good reason to not required cached render pipelines.

@cart @robtfm can you come up with anything?

**Comment by superdump** [comment]
Oh, and this needs rebasing after the depth/normal prepass PR was merged.

File: crates/bevy_ecs/src/schedule_v3/executor/multi_threaded.rs
Comment (kurtkuehnert) [reply crates/bevy_ecs/src/schedule_v3/executor/multi_threaded.rs:188]: fixed

**Comment by robtfm** [comment]
i think if you want a render pipeline it makes sense that it is always via a CachedRenderPipelineId. but are there scenarios where you would want a compute pipeline, or possibly no pipeline and just some buffer copying or something?

in my experience such things usually bypass phase items, but that could just be because i'm not making the most of the abstractions.

**Comment by kurtkuehnert** [comment]
From my understanding, the `RenderPhase` is designed to group/sort/batch together entities (`PhaseItem`s) for a rendering using a `TrackedRenderPass`. All preparation e.g buffer uploads, compute shaders, etc. should happen before that in the `Prepare` stage, or other parts of the `RenderGraph`.

To render an item a `RenderPipeline` is always required and storing a cached pipeline ID on the item seems ubiquitous and logical.

**Comment by robtfm** [comment]
for a `RenderPhase`, yes. i'm just pointing out we give up the ability to have `ComputePhase` or `CopyPhase` or anything like that.

**Comment by kurtkuehnert** [comment]
Right, but would they benefit from sorting/grouping/batching? 
I think they would need a separate abstraction since the draw function / render command ID is already necessary for all phase items.

**Comment by robtfm** [comment]
> Right, but would they benefit from sorting/grouping/batching?

that's the fundamental question. i could imagine for example a gpu-based physics library wanting to batch and compute-pass on islands. whether that would be useful, or if there are already too many other render-specific assumptions that make it impractical, i don't know.

> I think they would need a separate abstraction since the draw function / render command ID is already necessary for all phase items.

yes sorry you're right. because `PhaseItem` has `draw_function()`, and `Draw<T>` uses `TrackedRenderPass`, the phase items are inextricably tied to rendering. so as it stands there's no benefit to having `PhaseItem` and `CachedRenderPhaseItem` separately. 

so we could merge them, or we could preserve some theoretical non-render potential by moving `draw_function()` down to `CachedRenderPhaseItem` ... i think that might be the better option but maybe it's just needless complexity.

**Comment by superdump** [comment]
We could also merge it and change it if/when a need comes up.

**Comment by kurtkuehnert** [comment]
Good point. @robtfm 

I do agree with you @superdump.
If you do not restrict any current use cases, this will clean up the API quite nicely.
I want the render phase module to be as easy as possible to comprehend and adding all of these nuances, that might or might not be required, makes learning about rendering even more challenging.

If we decide that there is value in using Phases for compute as well, then we can consider building a separate abstraction or make `RenderPhase` more general purpose.

**Comment by cart** [comment]
I'm on board for the move toward simplification / more opinionated apis and we can then break out when the need arises.

**Comment by superdump** [comment]
One thing I realised while hacking on batching was that we may want to store the actual pipeline id on the entity, instead of (or in addition to) the cached pipeline id. Then we can avoid the hash map lookup when binding a pipeline.

**Comment by kurtkuehnert** [comment]
Ahh, that seems like a valid use case. Should I close this then?

**Comment by james7132** [comment]
> One thing I realised while hacking on batching was that we may want to store the actual pipeline id on the entity, instead of (or in addition to) the cached pipeline id. Then we can avoid the hash map lookup when binding a pipeline.

Is there a reason why we keep both? Seems like they're just two different ways to refer to the same pipeline. You can make a RenderPipeline outside of the cache, but the RenderPipelineId still always uniquely refers a pipeline.

**Comment by kurtkuehnert** [comment]
The `CachedPipelineId`s are tightly packed to look up in an array. If we were to use the `RenderPipelineId` then we would have to use a hash map lookup instead. Might be worth it though.

**Comment by james7132** [comment]
> The `CachedPipelineId`s are tightly packed to look up in an array. If we were to use the `RenderPipelineId` then we would have to use a hash map lookup instead. Might be worth it though.

That or we could take a sparse array approach, which should have the same performance characteristics and avoids the hash map lookup.

**Comment by kurtkuehnert** [comment]
Mhh good idea. I will give this a try later.

**Comment by kurtkuehnert** [comment]
@superdump @james7132 

From my understanding, we store **all** of our `RenderPipeline`s / `ComputePipeline`s in the `PipelineCache`.
They can either be queued directly using the `queue_pipeline` methods or created as part of specialized render/compute pipelines (which queue them if they are not already cached).

To access them the pipeline cache is always involved. We use the `get_pipeline` methods to look them up by their `CachedPipelineId` in a simple vector (not a hash map).

Rob, how would you look up a pipeline by their `RenderPipelineId` and which hash map lookup would this prevent? 
To me, it seems that this is not currently possible.
So I believe that this PR should be fine.
In theory, we could also clone our pipelines to avoid repetitive lookups in the pipeline cache.

James, I have tried using the `RenderPipelineId`s and `ComputePipelineId`s directly. Currently this does not work, because the id is only generated when the pipeline is created from the device. But the cache already stores the state of pipelines that need to be created in the future (i.e. pipelines, that wait on shaders to finish compiling).
We could special case pipeline ids and create them whenever a pipeline is queued in the cache, instead of when it is created. I think I will try this in another PR and see how this would play out. This change necessitates, that bevy pipelines can only be created using the `PipelineCache` (Note: that creating _raw_ pipelines using the underlying wgpu APIs is still possible to interop with third-party libraries).

**Comment by kurtkuehnert** [comment]
@james7132 #7459 does combine both IDs together.

**Pull request closed without merging**
